File: src/test/ironfist/path/astar/Node2D.java
/*
 * Created on Feb 10, 2005
 *
 */
package ironfist.path.astar;

import ironfist.loop.Level;
import ironfist.math.Vector2D;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * @author gremopm
 * 
 */
public class Node2D implements Node {

  private static final Vector2D[] DIRECTIONS = new Vector2D[]{
      Vector2D.get(1, 0),
      Vector2D.get(0, 1),
      Vector2D.get(-1, 0),
      Vector2D.get(0, -1)};

  private Level level;
  private Vector2D location;
  private Node2D parent;
  private double g;
  private double h;

  public Node2D(Level map, Vector2D location, Node2D parent) {
    this.level = map;
    this.location = location;
    this.parent = parent;
  }

  public Vector2D getLocation() {
    return location;
  }

  public Node getParent() {
    return parent;
  }

  public Collection<Node> getSuccessors() {
    List<Node> result = new ArrayList<Node>(DIRECTIONS.length);
    for (Vector2D direction : DIRECTIONS) {
      Vector2D position = location.add(direction);
      if ((parent == null || !position.equals(parent.getLocation()))
          && level.contains(position) && level.get(position) != null) {
        result.add(new Node2D(level, position, this));
      }
    }
    return result;
  }

  public void setCost(double g) {
    this.g = g;
  }

  public double getCost() {
    return g;
  }

  public double getTotal() {
    return g + h;
  }

  public double getEstimate() {
    return h;
  }

  public void setEstimate(double h) {
    this.h = h;
  }

  public boolean equals(Object obj) {
    return obj == null
        || (obj instanceof Node2D && ((Node2D) obj).getLocation()
          .equals(location));
  }

  public int hashCode() {
    return location.hashCode();
  }

  public String toString() {
    return location.toString();
  }
}

File: src/test/kotlin/ukalus/path/astar/Node2D.kt
/*
 * Created on Feb 10, 2005
 *
 */
package ukalus.path.astar

import ukalus.level.Level
import ukalus.math.Vector2D

/**
 * @author gremopm
 */
class Node2D(private val level: Level<*>, val location: Vector2D, override val parent: Node2D?) : Node {
    override var cost: Double = 0.0
    override var estimate: Double = 0.0

    override val successors: Collection<Node>
        get() = DIRECTIONS
                .map { location + it }
                .filter { (parent == null || it != parent.location) && level.contains(it) }
                .map { Node2D(level, it, this) }

    override val total
        get() = cost + estimate

    override fun equals(other: Any?) = other == null || other is Node2D && other.location == location

    override fun hashCode() = location.hashCode()

    override fun toString() = location.toString()

    companion object {
        private val DIRECTIONS = arrayOf(Vector2D(1, 0), Vector2D(0, 1), Vector2D(-1, 0), Vector2D(0, -1))
    }
}